pkg/reconciler/controllers/remediator.go

```diff
@@ -0,0 +1,342 @@
+		currentRV = cache.DeletedResourceVersion
+	} else {
+		toRemediate = obj
+	}
+
+	// If the current ResourceVersion was last remediated, skip remediation because
+	// it should match the declared resource.
+	if currentRV == lastRV {
+		// If no fight errors, return directly and remove the remediate cache
+		if len(r.FightErrors()) == 0 {
+			r.RemoveObj(id)
+			return ctrl.Result{}, nil
+		}
+		// Even though remediation is skipped, it still needs to detect fight to
+		// report the fight status.
+		_, fightErr := r.ResolveFight(time.Now(), obj)
+		if fightErr == nil {
+			// No more fight for the object, remove it from the queue.
+			r.RemoveFightError(id)
+			r.RemoveObj(id)
+			return ctrl.Result{}, nil
+		}
+
+		r.AddFightError(id, fightErr)
+		klog.Errorf("fight for %s still exists in the remediator", id)
+		// Requeue the request to address the fight.
+		r.DoneObj(id, currentRV)
+		return ctrl.Result{}, nil
+	}
+
+	// If the current ResourceVersion doesn't match the last remediated one, it is
+	// a new remediation.
+	klog.V(3).Infof("Remediator processing object %q (generation: %d, resourceversion: %s)", id, obj.GetGeneration(), obj.GetResourceVersion())
+	_, remErr := r.remediate(r.ControllerCtx, id, toRemediate)
+	if remErr != nil {
+		// To debug the set of events we've missed, you may need to comment out this
+		// block. Specifically, this makes things smooth for production, but can
+		// hide bugs (for example, if we don't properly process delete events).
+		if remErr.Code() == syncerclient.ResourceConflictCode {
+			// This means our cached version of the object isn't the same as the one
+			// on the cluster. We need to refresh the cached version.
+			if refreshErr := r.refresh(r.ControllerCtx, obj); refreshErr != nil {
+				klog.Errorf("Unable to update cached version of %q: %v", id, refreshErr)
+			}
+		}
+		klog.Errorf("Failed to remediate %q: %v", id, remErr)
+		// The object is marked done in the queue so that it can be requeued next time.
+		// If there is no error, the next reconciliation will succeed and clear the object from the queue.
+		// If an error occurs, the next reconciliation will re-process the remediate request.
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(3).Infof("Successfully reconciled %q", id)
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager registers the Remediator Controller.
+func (r *Remediator) SetupWithManager(mgr ctrl.Manager) error {
+	// an event channel to trigger reconciliation
+	timerEvents := make(chan event.GenericEvent)
+
+	// run the timer as a goroutine to trigger the reconciliation periodically
+	go func() {
+		remediateTimer := time.NewTimer(time.Second)
+		defer remediateTimer.Stop()
+
+		for {
+			select {
+			case <-r.ControllerCtx.Done():
+				// The controller is shutting down, close the timerEvents channel and exit
+				close(timerEvents)
+				close(r.DoneCh) // inform finalizer the remediator is done
```

Comment:
The context cancel doesn't kill those threads, but those threads will fail because the context is canceled, and won't be requeued because the controller is shutting down. The downside is that there will be remediate errors in the logs because of the context cancelation.

We can pause the remediator (prevents new threads from being executed) and wait for the currently running thread (should be a single one) to complete before closing the doneCh (prevents errors from spamming the remediator log).

Reply:
done